keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,464 @@
+    
+### Publishing
+
+The `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` vendor extensions are published via `/openapi/v2` as is.
+
+### Suggested marker syntax
+
+In analogy to `+required`, `+optional` we propose to add a marker to kubebuilder's
+controller-gen named `+immutable`, which covers the whole subtree of the object 
+(compare field-selection section for extension).
+
+```
+// The name can not be changed after creation.
+// +immutable
+Name string
+
+// The list of containers can not change AT ALL after creation.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+Containers []Containers
+```
+
+The `x-kubernetes-immutable-keys: true` vendor extension is expressed via `// +immutable=keys` in Golang types:
+
+### Future outline sketch: native resources
+
+For native resources we add a pre-immutability-check normalization step for objects 
+decoded from JSON which have normalizations defined:
+
+1. versioned JSON blob comes in a request
+2. unmarshalled into versioned Golang struct
+3. defaulting
+4. conversion to internal
+5. if immutability is enabled for the resource:
+   1. marshalling into JSON in-memory
+   2. normalization creating a copy with shared data-structures
+   3. strict immutability check against the old object, coming from proto assuming it is normalized.
+
+### Future outline: protobuf
+
+Protobuf encoding unifies unset, empty and null values for slices and maps. Those three cases 
+cannot be differentiated an lead to the same on-the-wire value. In native types
+we use a semantic equality to factor those differences out if JSON is used on the wire, but
+protobuf is used for storage. In CRs decoding and encoding is always faithful with the 
+traditional JSON format used on-the-wire and for storage.
+
+When adding protobuf encoding to CRDs in the future, we have to (without major, non-standard
+efforts) identify unset, empty and null for CRs as well. This leads to the idea to use a less
+strict equality in the immutabiltiy checks with that case in mind. But protobuf encoding in
+native types also has a normalization effect, namely posted JSON object are normalized through
+the encoding to protobuf when writing to etcd. 
+
+Hence, it looks sensible to split normalization from immutability equality, and keep a strict 
+deep-equal equality even for protobuf, and potentially native types (if we decide to implement 
+immutability for those).
+
+With the container example we get this, with a sketch of a `+normalize` marker which
+normalizes empty and null to undefined:
+
+```
+// The list of containers can not change AT ALL after creation, modulo
+// empty, null and undefined.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+// +normalize=undefined
```

Comment:
I really strongly object to letting our deep-equal & proto round trip behavior leak into our api design specification. API authors should not have to understand go's peculiar deep equal behavior.